std/json.c3

```diff
@@ -0,0 +1,428 @@
+@require_once "std/arithmetic.c3"
+@require_once "std/string.c3"
+@require_once "std/memory.c3"
+@require_once "std/vector.c3"
+@require_once "std/wrappers.c3"
+@require_once "std/iterators.c3"
+
+typedef Cursor : {
+    data      : StringView,
+    position  : isize, // TODO should this be an iterator?
+    allocator : Allocator,  // FIXME use global allocator.
+}
+
+typedef JSON_Node : {
+    type     : int, // 0: null, 1: true, 2: false, 3: string, 4: number,
+                    // 5: array, 6: object
+    value    : Optional<StringView>,
+    children : Optional<Vector<JSON_Node>>,
+}
+
+// TODO something like new<JSON_Node> would be nice, but we don't support
+// function specializations.
+function new_json_node : (node_type : int) -> JSON_Node = {
+    return {
+        node_type,
+        make_optional<StringView>(),
+        make_optional<Vector<JSON_Node>>(),
+    };
+}
+
+function char : (literal : u8[1]&) -> u8 = {
+    return literal[0];
+}
+
+function is_eof : (cursor : Cursor&) -> bool = {
+    return cursor.position >= cursor.data.length;
+}
+
+function peak : (cursor : Cursor&) -> Optional<u8> = {
+    if !cursor:is_eof() {
+        return make_optional(cursor.data.data[cursor.position]);
+    }
+
+    return make_optional<u8>();
+}
+
+function next : (cursor : Cursor&) -> Optional<u8> = {
+    // TODO this should be const (can't use with has_value() right now).
+    let token : Optional<u8> = cursor:peak();
+
+    if token:has_value() {
+        cursor.position += 1;
+    }
+
+    return token;
+}
+
+function is_json_whitespace : (token : u8) -> bool = {
+    // JSON defines whitespace as one of { space, linefeed, carriage return,
+    // horizontal tab }. If we use string literals instead of code points, then
+    // our code doesn't care about the underlying encoding :).
+    return token == char(" ") or token == char("\n")
+           or token == char("\r") or token == char("\t");
+}
+
+function eat : (cursor : Cursor&, token : u8) -> bool = {
+    // TODO this should be const (can't use with has_value() right now).
+    let next_char : Optional<u8> = cursor:peak();
+
+    let next_char_data : u8 = next_char:data();
+    if next_char:has_value() and next_char_data == token {
+        cursor:next();
+        return true;
+    }
+
+    return false;
+}
+
+function eat : (cursor : Cursor&, string : StringView) -> bool = {
+    for token in string:get_iter() {
+        if !cursor:eat(token) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function eat_whitespace : (cursor: Cursor&) -> void = {
```

Comment:
I kinda like the `eat_*` convention. Might steal that next time I'm working on the parser

Reply:
I stole it from someone else myself :)